File: slices.cpp
//
// Slices
// UCF Online High School Programming Contest 2019
// Jim Geist jimg@knights.ucf.edu
//
#include <cstdio>
#include <cstring>

int main()
{
    int cases;
    scanf("%d", &cases);

    while (cases--) {
	int n, k;
	scanf("%d %d", &n, &k);

	int jim = 0, kelly = 0;
	bool jimHates = false, kellyHates = false;

	char name[10];
	while (n--) {
	    int level;
	    
	    scanf("%s %d", name, &level);
	    if (strcmp(name, "Jim") == 0) {
		jim = level;
	    } else {
		kelly = level;
	    }

	    if (jim > kelly + k) {
		kellyHates = true;
	    }

	    if (kelly > jim + k) {
		jimHates = true;
	    }
	}

	if (jimHates == false && kellyHates == false) {
	    printf("Everything is good\n");
	} else if (jimHates == false && kellyHates == true) {
	    printf("Kelly hates Jim\n");
	} else if (jimHates == true && kellyHates == false) {
	    printf("Jim hates Kelly\n");
	} else if (jimHates == true && kellyHates == true) {
	    printf("Their friendship is doomed\n");
	}	
    }    
}

File: slices.py
#solution to Slices by David Harmeyer

#Current issues:
#samples don't match the problem description. Problem statement says it is okay
#to be one away, samples say oppisite
#Weak data: only one case tests off-by-one errors, and it is a border case.
#someone could easily get part of it wrong, but start their counters at level 1,
#and get it correct when it should be incorrect.
T=int(input());

for tt in range(0, T):
  n,k=map(int, input().split());
  eHatesT=False;
  tHatesE=False;
  e,t=0, 0;
  for i in range(0, n):
    solver=input().split()[0];
    if (solver=="Kelly"):
      #what kind of self-respecting programming language doesn't have the ++ operator?!
      e+=1;
    else:
      t+=1;
    if e>t+k:
      tHatesE=True;
    if t>e+k:
      eHatesT=True;
  if eHatesT and tHatesE:
    print("Their friendship is doomed");
  elif eHatesT:
    print("Kelly hates Jim");
  elif tHatesE:
    print("Jim hates Kelly");
  else:
    print("Everything is good");
    



